Extend rxm and coll providers to be used with offload provider

grom72 opened the pull request:
Extend rxm and coll providers to work with collective offload providers:
- collective offload provider is given via FI_OFFLOAD_COLL_PROVIDER environment variable
- multinode tests do not fail if a provider does not explicitly support the requested collective operation
- collective offload provider's capabilities override the util coll provider's capabilities

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/libfabric/2)
<!-- Reviewable:end -->

Commit:
Intel-CI: Adding code in Jenkinsfile for creating a logdir for daos

add call to create a logdir for daos on jfcst-daos during build stage
added a parallel summary stage.

Signed-off-by: Nikhil Nanal <nikhil.nanal@intel.com>

Commit:
Intel-CI: making change to the tests.py for generating daos summary

modifying run command to generate logs for daos.

Signed-off-by: Nikhil Nanal <nikhil.nanal@intel.com>

Commit:
Intel CI: Summary.py updated for generating daos summary

Added class  to summarize daos logs. modified check_name,
check_pass, check_fail and check_line methods

Signed-off-by: Nikhil Nanal <nikhil.nanal@intel.com>

jtamzn committed:
prov/efa: Add more tracepoints

Signed-off-by: Tang, Jingyin <jytang@amazon.com>

grom72 committed:
prov/coll: Prepare coll_cq to be used in by collective providers

coll_cq implementation can be reused by other collective providers.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: Collective offload capabilities reported with highiest priority

Collective offload capabilities reported if offload provider is available
otherwise util collective provider capabilities are reported.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: Collective offload fabric setup integrated into rxm fabric initialization

It is rxm provider responsability to initialize collective offload provider's fabric.
Otherwise collective offload functionality will not be available

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
core/fabric: Collective offload provider set via FI_OFFLOAD_COLL_PROVIDER

FI_OFFLOAD_PROVIDER environment variable shall be set to offload provider name
to instruct libcabric to setup and use particular provider.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: Create eq for collective offload provider

Peer provider must create peer_eq for offload provider, to allow offload provider
reporting events to peer provider.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

ldorau reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/2)* status: 0 of 9 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

grom72 committed:
prov/rxm: FI_PEER flag force fi_join() with util_coll_ep

Offload provider may execute collective operations via util_coll provider.
It must call fi_join() operation to get struct mc required for collective operations.
It can only call fi_join() on it's peer provider (e.g. rxm). FI_PEER flag is used
to inform peer provider to coll fi_join() operation for util_coll_ep

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

j-xiong committed:
prov/rxm: Always use rendezvous protocol for ZE device memory send

Bounce buffer copy overhead is high for ZE device memory. The rendezvous
protocol takes advantage of GPU RDMA and performs better even for small
messages.

Signed-off-by: Jianxin Xiong <jianxin.xiong@intel.com>

Commit:
fabric: Remove FI_PEER_DOMAIN

The FI_PEER flag is sufficient for allocating peer objects.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

Commit:
prov/net: Remove unused ep::hit_cnt field

This is carry over from the tcp provider but unused.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

vidsouza committed:
prov/efa: Do no ignore hints->domain_attr->name

Prior to this patch, the provider would ignore the domain
name in hints->domain_attr->name even if this was set.
This patch fixed this issue.

Signed-off-by: Vishwas Dsouza <vidsouza@amazon.com>

vidsouza committed:
fabtests/pytest/efa: Add efa device selection test to fabtests suite

This test tests the -d <domain_name> argument of the fabtests.
All the EFA domain names are tested with this test with the
corresponding fabtest.

Signed-off-by: Vishwas Dsouza <vidsouza@amazon.com>

grom72 committed:
prov/rxm: Set offload_coll_mask based on actual configuration

offload_coll_mask value is calculated based on the actual offload capabilities
confirmed by fi_query_collective().

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

haichangsi reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/2)* status: 0 of 14 files reviewed, all discussions resolved (waiting on @ldorau)



<!-- Sent from Reviewable.io -->

ldorau reviewed (changes requested):
> Reviewed 4 of 13 files at r4, 1 of 1 files at r5, 8 of 9 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/2)* status: all files reviewed, 5 unresolved discussions (waiting on @grom72)

___
*[`prov/rxm/src/rxm_ep.c` line 393 at r7](https://reviewable.io/reviews/pmem/libfabric/2#-NIXMpMU62wnt_blEBaW:-NIXMpMU62wnt_blEBaX:bsnnk5t) ([raw file](https://github.com/pmem/libfabric/blob/4fb844ec8273b0a795646fcafea2ffda7a1d893a/prov/rxm/src/rxm_ep.c#L393)):*
> ```C
> 	rxm_ep = container_of(ep, struct rxm_ep, util_ep.ep_fid);
> 
> 	//FI_PEER flag is used to force util_coll context 
> ```

Redundant space at the end



_Code quote:_
```C
t·
```
___
*[`prov/rxm/src/rxm_ep.c` line 393 at r7](https://reviewable.io/reviews/pmem/libfabric/2#-NIXMsaQ9yiboUX_8BD5:-NIXMsaQ9yiboUX_8BD6:b-8nk0v8) ([raw file](https://github.com/pmem/libfabric/blob/4fb844ec8273b0a795646fcafea2ffda7a1d893a/prov/rxm/src/rxm_ep.c#L393)):*
> ```C
> 	rxm_ep = container_of(ep, struct rxm_ep, util_ep.ep_fid);
> 
> 	//FI_PEER flag is used to force util_coll context 
> ```

Use `/* */` C-style comment, please



_Code quote:_
```C
	//FI_PEER flag is used to force util_coll context
	//where fi_join() is called from offload provider
```
___
*[`prov/rxm/src/rxm_ep.c` line 402 at r7](https://reviewable.io/reviews/pmem/libfabric/2#-NIXMzhp-JpO926BB2Cv:-NIXMzhp-JpO926BB2Cw:b3kqx) ([raw file](https://github.com/pmem/libfabric/blob/4fb844ec8273b0a795646fcafea2ffda7a1d893a/prov/rxm/src/rxm_ep.c#L402)):*
> ```C
> 		if (ret)
> 			goto err_util_coll;
> 		// It is collective offload provider responsibility to store util_coll provider mc
> ```

.



_Code quote:_
```C
// It is collective offload provider responsibility to store util_coll provider mc
```
___
*[`prov/rxm/src/rxm_msg.c` line 818 at r7](https://reviewable.io/reviews/pmem/libfabric/2#-NIXNA0F7GfNoEdPBpFA:-NIXNA0F7GfNoEdPBpFB:bbeo5zj) ([raw file](https://github.com/pmem/libfabric/blob/4fb844ec8273b0a795646fcafea2ffda7a1d893a/prov/rxm/src/rxm_msg.c#L818)):*
> ```C
> 				   rxm_ep_sar_calc_segs_cnt(rxm_ep, data_len));
> 	} else {
> rndv_send:
> ```

An empty line should be used before a label, not after.



_Code quote:_
```C
	} else {
rndv_send:

		ret = rxm_alloc_rndv_buf(rxm_ep, rxm_conn, context,
```
___
*[`src/fabric.c` line 841 at r7](https://reviewable.io/reviews/pmem/libfabric/2#-NIXNIqzAIkAjGAvTwni:-NIXNIr-4shNmJz6qqjy:b-o3lwxt) ([raw file](https://github.com/pmem/libfabric/blob/4fb844ec8273b0a795646fcafea2ffda7a1d893a/src/fabric.c#L841)):*
> ```C
> 
> 	fi_param_define(NULL, "offload_coll_provider", FI_PARAM_STRING,
> 			"The name of colective offload provider (default: empty - no provider)");
> ```

`of a collective offload provider`



_Code quote:_
```C
of colective offload provider
```


<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/2)* status: all files reviewed, 5 unresolved discussions (waiting on @grom72)



<!-- Sent from Reviewable.io -->

aingerson committed:
fabtests/runfabtests.sh: add option to run with gdb to get backtraces

Signed-off-by: Alexia Ingerson <alexia.ingerson@intel.com>

Commit:
prov/net: Create union for msg headers

Define an explicit union for message headers and replace
embedded unions defined in structures.  This will allow
passing the union to match calls.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

Commit:
prov/net: Change xnet_check_match to use union xnet_hdrs

Replace the active_rx parameter with a pointer to the headers,
which is all that's needed.  This separates match searching
from the active_rx.

Carry through this change up the call stack.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

Commit:
prov/net: Fix memory leak in peek/claim path

Fix assertion when destroying buffer pools that not all xfer_entry's
have been returned.  Move freeing of xfer_entry to immediately after
writing a CQ entry.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

ldorau committed:
.gitignore: Add configure~ to .gitignore

Add configure~ to .gitignore.

Signed-off-by: Lukasz Dorau <lukasz.dorau@intel.com>

zachdworkin committed:
Merge pull request #8117 from nikhilnanal/daos_summary

Daos summary

grom72 committed:
prov/rxm: Report only coll offload capabilities when OFI_OFFLOAD_PROV_ONLY is set

fi_query_collective() reports only collective offload provider capabilities if
OFI_OFFLOAD_PROV_ONLY flag is set. Otherwise the sum of both providers' capabilities
is reported.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

wzamazon committed:
prov/efa: add a FI_WARN about endpoint address after it is created

This patch added a call to FI_WARN to print endpoint's address
and libfabric version after endpoint is created.

Signed-off-by: Wei Zhang <wzam@amazon.com>

j-xiong committed:
fabtests: Fix release package path in README.md

fabtests release packages are co-located with libfabric since 1.7.0.

Signed-off-by: Jianxin Xiong <jianxin.xiong@intel.com>

j-xiong committed:
prov/rxm: Fix incorrect use of OFI_UNUSED()

The macro is for function parameters only.

Signed-off-by: Jianxin Xiong <jianxin.xiong@intel.com>

zachdworkin committed:
contrib/intel/jenkins: Make git clone's specifiy which branch

In the event where git cloning upstream/main provides newer python files
than the Jenkinsfile is expecting. There will be failures when the new
python expected the old Jenkinsfile to call something that it didn't.
Changing to checkout the same branches python files to avoid this mismatch.

Signed-off-by: Zach Dworkin <zachary.dworkin@intel.com>
